File: app/aem/core/src/main/java/com/cognifide/apm/core/endpoints/ScriptRemoveServlet.java
/*-
 * ========================LICENSE_START=================================
 * AEM Permission Management
 * %%
 * Copyright (C) 2013 Wunderman Thompson Technology
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
package com.cognifide.apm.core.endpoints;

import com.cognifide.apm.api.scripts.Script;
import com.cognifide.apm.api.services.ScriptFinder;
import com.cognifide.apm.core.Property;
import com.cognifide.apm.core.scripts.ScriptStorage;
import com.cognifide.apm.core.utils.ServletUtils;
import java.io.IOException;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import javax.jcr.RepositoryException;
import javax.servlet.Servlet;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletResponse;
import org.apache.commons.lang.StringUtils;
import org.apache.sling.api.SlingHttpServletRequest;
import org.apache.sling.api.SlingHttpServletResponse;
import org.apache.sling.api.resource.ResourceResolver;
import org.apache.sling.api.servlets.HttpConstants;
import org.apache.sling.api.servlets.SlingAllMethodsServlet;
import org.osgi.service.component.annotations.Component;
import org.osgi.service.component.annotations.Reference;

@Component(
		immediate = true,
		service = Servlet.class,
		property = {
				Property.PATH + "/bin/cqsm/remove",
				Property.METHOD + HttpConstants.METHOD_POST,
				Property.DESCRIPTION + "CQSM Remove Scripts Servlet",
				Property.VENDOR
		}
)
public class ScriptRemoveServlet extends SlingAllMethodsServlet {

	@Reference
	private transient ScriptStorage scriptStorage;

	@Reference
	private transient ScriptFinder scriptFinder;

	@Override
	protected void doPost(SlingHttpServletRequest request, SlingHttpServletResponse response)
			throws ServletException, IOException {
		final String all = request.getParameter("confirmation");
		final String fileName = request.getParameter("file");
		ResourceResolver resolver = request.getResourceResolver();
		if (fileName != null) {
			removeSingleFile(resolver, response, fileName);
		} else if (all != null) {
			removeAllFiles(resolver, response, all);
		} else {
			response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
			ServletUtils.writeMessage(response, ServletUtils.ERROR_RESPONSE_TYPE, "Invalid arguments specified");
		}
	}

	private void removeAllFiles(ResourceResolver resolver, SlingHttpServletResponse response, String all)
			throws IOException {
		if (!Boolean.parseBoolean(all)) {
			ServletUtils.writeMessage(response, ServletUtils.ERROR_RESPONSE_TYPE, "Remove all scripts is not confirmed");
		} else {
			final List<String> paths = new LinkedList<>();
			final List<Script> scripts = scriptFinder.findAll(resolver);

			try {
				for (Script script : scripts) {
					final String path = script.getPath();

					scriptStorage.remove(script, resolver);
					paths.add(path);
				}

				final Map<String, Object> context = new HashMap<>();
				context.put("paths", paths);

				ServletUtils.writeMessage(response, ServletUtils.SUCCESS_RESPONSE_TYPE,
						String.format("Scripts removed successfully, total: %d", scripts.size()), context);
			} catch (RepositoryException e) {
				response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
				ServletUtils.writeJson(response,
						"Cannot save remove all scripts. Repository error: " + e.getMessage());
			}
		}
	}

	private void removeSingleFile(ResourceResolver resolver, SlingHttpServletResponse response,
			String fileName) throws IOException {
		if (StringUtils.isEmpty(fileName)) {
			ServletUtils.writeMessage(response, ServletUtils.ERROR_RESPONSE_TYPE, "File name to be removed cannot be empty");
		} else {
			final Script script = scriptFinder.find(fileName, resolver);
			if (script == null) {
				ServletUtils
						.writeMessage(response, ServletUtils.ERROR_RESPONSE_TYPE, String.format("Script not found: '%s'", fileName));
			} else {
				final String scriptPath = script.getPath();

				try {
					scriptStorage.remove(script, resolver);

					ServletUtils.writeMessage(response, ServletUtils.SUCCESS_RESPONSE_TYPE,
							String.format("Script removed successfully: %s", scriptPath));
				} catch (RepositoryException e) {
					response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
					ServletUtils.writeJson(response,
							String.format("Cannot remove script: '%s'." + " Repository error: %s", scriptPath,
									e.getMessage()));
				}
			}
		}
	}
}

File: app/aem/core/src/main/kotlin/com/cognifide/apm/core/endpoints/ScriptRemoveServlet.kt
package com.cognifide.apm.core.endpoints

import org.apache.sling.api.servlets.SlingAllMethodsServlet
import org.osgi.service.component.annotations.Component
import org.osgi.service.component.annotations.Reference
import javax.servlet.Servlet
import com.cognifide.apm.core.scripts.ScriptStorage
import com.cognifide.apm.api.services.ScriptFinder
import org.apache.sling.api.SlingHttpServletRequest
import org.apache.sling.api.SlingHttpServletResponse
import javax.servlet.http.HttpServletResponse
import com.cognifide.apm.core.utils.ServletUtils
import org.apache.commons.lang.StringUtils
import org.apache.sling.api.resource.ResourceResolver
import javax.jcr.RepositoryException
import com.cognifide.apm.core.Property

@Component(
        immediate = true,
        service = [Servlet::class],
        property = [
            Property.PATH + "/bin/apm/remove",
            Property.METHOD + "POST",
            Property.DESCRIPTION + "APM Remove Scripts Servlet",
            Property.VENDOR
        ])

class ScriptRemoveServlet : SlingAllMethodsServlet() {

    @Reference
    @Transient
    private lateinit var scriptStorage: ScriptStorage

    @Reference
    @Transient
    private lateinit var scriptFinder: ScriptFinder

    override fun doPost(request: SlingHttpServletRequest, response: SlingHttpServletResponse) {
        val all = request.getParameter("confirmation")
        val fileName = request.getParameter("file")
        val resourceResolver = request.resourceResolver

        if (fileName != null) {
            removeSingleFile(resourceResolver, response, fileName)
        } else if (all != null) {
            removeAllFiles(resourceResolver, response, all)
        } else {
            response.setStatus(HttpServletResponse.SC_BAD_REQUEST)
            ServletUtils.writeMessage(response, "error", "Invalid arguments specified")
        }
    }

    private fun removeSingleFile(resourceResolver: ResourceResolver, response: SlingHttpServletResponse, fileName: String) {
        if (StringUtils.isEmpty(fileName)) {
            ServletUtils.writeMessage(response, "error", "File name to be removed cannot be empty")
        } else {
            val script = scriptFinder.find(fileName, resourceResolver)
            if (script == null) {
                ServletUtils.writeMessage(response, "error", "Script not found: $fileName")
            } else {
                val scriptPath = script.path

                try {
                    scriptStorage.remove(script, resourceResolver)
                    ServletUtils.writeMessage(response, "success", "Script removed successfully: $scriptPath")
                } catch (e: RepositoryException) {
                    response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
                    ServletUtils.writeMessage(response, "error", "Cannot remove script: $scriptPath. Repository error: ${e.message}")
                }
            }
        }
    }

    private fun removeAllFiles(resourceResolver: ResourceResolver, response: SlingHttpServletResponse, all: String) {
        //todo Kotlin way?
        if (!java.lang.Boolean.parseBoolean(all)) {
            ServletUtils.writeMessage(response, "error", "Remove all scripts is not confirmed")
        } else {
            var paths: MutableList<String> = mutableListOf()
            val scripts = scriptFinder.findAll(resourceResolver)

            try {
                for (script in scripts) {
                    val path = script.path
                    scriptStorage.remove(script, resourceResolver)
                    paths.add(path)
                }
                val context: MutableMap<String, Any> = hashMapOf()
                context.put("paths", paths)
                ServletUtils.writeMessage(response, "success", "Scripts removed successfully, total: ${scripts.size}", context)
            } catch (e: RepositoryException) {
                response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
                ServletUtils.writeMessage(response, "error", "Cannot save scripts removal. Repository error: ${e.message}")
            }
        }
    }
}